packages/editor-ui/src/modules/tags.ts

```diff
@@ -0,0 +1,103 @@
+
+import { ActionContext, Module } from 'vuex';
+import {
+	ITag,
+	ITagsState,
+	IRootState,
+} from '../Interface';
+import { createTag, deleteTag, getTags, updateTag } from '../api/tags';
+import Vue from 'vue';
+
+const module: Module<ITagsState, IRootState> = {
+	namespaced: true,
+	state: {
+		tags: {},
+		isLoading: false,
+		fetchedAll: false,
+		fetchedUsageCount: false,
+	},
+	mutations: {
+		setLoading: (state: ITagsState, isLoading: boolean) => {
+			state.isLoading = isLoading;
+		},
+		setAllTags: (state: ITagsState, tags: ITag[]) => {
+			state.tags = tags
+				.reduce((accu: {[id: string]: ITag}, tag: ITag) => {
+					accu[tag.id] = tag;
+
+					return accu;
+				}, {});
+			state.fetchedAll = true;
+		},
+		upsertTags(state: ITagsState, tags: ITag[]) {
+			tags.map((tag) => {
+				if (state.tags[tag.id]) {
+					const newTag = {
+						...state.tags[tag.id],
+						...tag,
+					};
+					Vue.set(state.tags, tag.id, newTag);
+				}
+				else {
+					Vue.set(state.tags, tag.id, tag);
+				}
+			});
+		},
+		deleteTag(state: ITagsState, id: string) {
+			Vue.delete(state.tags, id);
+		},
+	},
+	getters: {
+		tags(state: ITagsState): ITag[] {
+			return Object.values(state.tags)
+				.sort((a: ITag, b: ITag) => a.name.localeCompare(b.name));
+		},
+		isLoading: (state: ITagsState): boolean => {
+			return state.isLoading;
+		},
+		hasTags: (state: ITagsState): boolean => {
+			return Object.keys(state.tags).length > 0;
+		},
+		getTagById: (state: ITagsState) => {
+			return (id: string) => state.tags[id];
+		},
+	},
+	actions: {
+		fetchAll: async (context: ActionContext<ITagsState, IRootState>, params: {force?: boolean, withUsageCount?: boolean}) => {
+			const {force, withUsageCount} = params || {};
+			if (!force && context.state.fetchedAll && context.state.fetchedUsageCount === !!withUsageCount) {
```

Comment:
Can we think of a better way of ensuring that a possibly undefined arg becomes a boolean?

Reply:
could use `Boolean(withUsageCount)`? is that what you mean?

Reply:
Not a fan of optional booleans (union of three states) and of indirect casting with `!!`.

Since `fetchedUsageCount` is a boolean, the comparison to `withUsageCount` will work the same when it is `false` and when it is `undefined`, so `!!` is unneeded.

Or a more type-safe way would be to set defaults during destructuring, if they fit with your logic:

```
fetchAll: async (context: ActionContext<ITagsState, IRootState>, { force = false, withUsageCount = false }: { force: boolean, withUsageCount: boolean })
```

This would also simplify lines 67 and 73.

Reply:
looks good